doc/architectural/CBMC-developer-guide.md

```diff
@@ -1,11 +1,196 @@
+internals of the system and how to get started on development.
+
+## CMake files ##
+
+FIXME
+
+## Personal configuration: config.inc, macro DEBUG ##
+
+FIXME
+
+## Generating doxygen documentation ##
+
+Run `doxygen` in `/src`. HTML output will be created in `/doc/html`. The
+index page is `/doc/html/index.html`.
+
+Apart from the (user-orientated) CBMC user manual and this document, most
+of the rest of the documentation is inline in the code as `doxygen` and
+some comments. A man page for CBMC, goto-cc and goto-instrument is
+contained in the `doc/` directory and gives some options for these
+tools. All of these could be improved and patches are very welcome. In
+some cases the algorithms used are described in the relevant papers.
+
+## Coding standards ##
+
+See <a
+href="https://github.com/diffblue/cbmc/blob/master/CODING_STANDARD.md">
+`CODING_STANDARD.md`</a> file in the root of the CBMC repository.
+
+CPROVER is written in a fairly minimalist subset of C++; templates and
+meta-programming are avoided except where necessary.
+External library dependencies are avoided (only STL and a SAT solver
+are required). Boost is not used. The `util` directory contains many
+utilities that are not (yet) in the standard library.
+
+Patches should be formatted so that code is indented with two space
+characters, not tab and wrapped to 80 columns. Headers for doxygen
+should be given (and preferably filled!) and the author will be the
+person who first created the file. Add doxygen comments to
+undocumented functions as you touch them. Coding standards
+and doxygen comments are enforced by CI before a patch can be
+merged by running `clang-format` and `cpplint`.
+
+Identifiers should be lower case with underscores to separate words.
+Types (classes, structures and typedefs) names must end with a `t`.
+Types that model types (i.e. C types in the program that is being
+interpreted) are named with `_typet`. For example `ui_message_handlert`
+rather than `UI_message_handlert` or `UIMessageHandler` and
+`union_typet`.
+
+
+\section section-background-concepts Background Concepts
+
+## AST: types, globals, variables, functions, code blocks, language primitives, assignments, expressions, variables ##
+
+FIXME
+
+## CFG ##
+
+FIXME
+
+## Bounded model checking (from the CBMC manual) ##
+
+FIXME
+
+### SSA ###
+
+FIXME
 
-\author Martin Brain, Peter Schrammel
+## SAT and SMT ##
 
-\section overview-dirs Overview of CPROVER Directories
+FIXME
 
-## `src/`
+## Static analysis ##
+
+FIXME
+
+
+\section section-cprover-architecture CPROVER Architecture
+
+This section provides a graphical overview of how CBMC fits together.
+CBMC takes C code or a goto-binary as input and tries to emit traces
+of executions that lead to crashes or undefined behaviour. The diagram
+below shows the intermediate steps in this process.
+
+\dot
+digraph G {
+
+  rankdir="TB";
+  node [shape=box, fontcolor=blue];
+
+  subgraph top {
+    rank=same;
+    1 -> 2 -> 3 -> 4;
```

Comment:
You can give these informative short names instead of 1, 2, 3, ...

Reply:
I didn't write this. It seems clear enough to me.